core/src/main/java/kafka/server/share/DelayedShareFetch.java

```diff
@@ -149,15 +149,27 @@ public void onComplete() {
      */
     @Override
     public boolean tryComplete() {
-        log.trace("Try to complete the delayed share fetch request for group {}, member {}, topic partitions {}",
-            shareFetchData.groupId(), shareFetchData.memberId(),
-            shareFetchData.partitionMaxBytes().keySet());
+        // There can be multiple threads which might invoke tryComplete for same share fetch request
+        // hence check if delay share fetch request is already completed. If yes, return true.
+        // However, this check alone cannot guarantee that request is really completed. It is possible that
+        // tryComplete is invoked by multiple threads and state has yet not updated. Hence, we need to check
+        // the forceComplete response as well.
+        if (isCompleted()) {
+            return true;
+        }
 
         topicPartitionDataFromTryComplete = acquirablePartitions();
 
-        if (!topicPartitionDataFromTryComplete.isEmpty())
-            return forceComplete();
-        log.info("Can't acquire records for any partition in the share fetch request for group {}, member {}, " +
+        if (!topicPartitionDataFromTryComplete.isEmpty()) {
+            boolean result = forceComplete();
+            // If invocation of forceComplete is not successful, then that means the request is already completed
+            // hence release the acquired locks.
+            if (!result) {
+                releasePartitionLocks(shareFetchData.groupId(), topicPartitionDataFromTryComplete.keySet());
+            }
+            return result;
+        }
+        log.trace("Can't acquire records for any partition in the share fetch request for group {}, member {}, " +
```

Comment:
The frequency of this log line is phenomenal. I wonder whether it's really helpful or just likely to flood the logs to the extent that it's impossible to see anything else.

Reply:
yeah, I think I made it INFO at the beginning because I was testing the purgatory stuff, but going forward we will make the logs as trace/debug. I forgot to change it to DEBUG when it got merged.

Reply:
I have removed this log line, we can check the purgatory metric to see waiting requests.

Reply:
Thanks. That seems much more appropriate.